File: bot/domain/valueobjects/song_metadata.py


from __future__ import annotations
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, Dict, Any, List

from ..valueobjects.source_type import SourceType
from ..valueobjects.song_status import SongStatus
@dataclass(frozen=True)
class SongMetadata:
    """Immutable song metadata"""
    title: str
    artist: str
    duration: int  # seconds
    album: Optional[str] = None
    thumbnail_url: Optional[str] = None
    release_date: Optional[str] = None
    genres: List[str] = field(default_factory=list)

    @property
    def display_name(self) -> str:
        """Human-readable song name"""
        if self.artist and self.title:
            return f"{self.artist} - {self.title}"
        return self.title

    @property
    def duration_formatted(self) -> str:
        """Duration in MM:SS format"""
        if self.duration <= 0:
            return "00:00"
        minutes, seconds = divmod(self.duration, 60)
        return f"{minutes:02d}:{seconds:02d}"

    @property
    def search_query(self) -> str:
        """Generate search query for finding this song"""
        parts = []
        if self.artist:
            parts.append(self.artist)
        if self.title:
            parts.append(self.title)
        return " ".join(parts)

File: internal/domain/valueobjects/song_metadata.go
package valueobjects

import "fmt"

// SongMetadata contains metadata information about a song
type SongMetadata struct {
	Title     string `json:"title"`
	Artist    string `json:"artist,omitempty"`
	Duration  int    `json:"duration"` // seconds
	Thumbnail string `json:"thumbnail,omitempty"`
	Uploader  string `json:"uploader,omitempty"`
}

// DisplayName returns the best display name for the song
func (m *SongMetadata) DisplayName() string {
	if m.Artist != "" {
		return fmt.Sprintf("%s - %s", m.Artist, m.Title)
	}
	return m.Title
}

// DurationFormatted returns duration in MM:SS format
func (m *SongMetadata) DurationFormatted() string {
	if m.Duration <= 0 {
		return "00:00"
	}

	minutes := m.Duration / 60
	seconds := m.Duration % 60
	return fmt.Sprintf("%02d:%02d", minutes, seconds)
}
